rust-runtime/aws-smithy-types/src/type_erasure.rs

```diff
@@ -101,9 +102,11 @@ impl<T: fmt::Debug + Send + Sync + 'static> DerefMut for TypedBox<T> {
 pub struct TypeErasedBox {
     field: Box<dyn Any + Send + Sync>,
     #[allow(clippy::type_complexity)]
-    debug: Box<
+    debug: Arc<
```

Comment:
Is it possible to to create a struct with a PhantomData to do the casting instead of using an Arc? Something like:
```rust
#[derive(Debug, Clone)]
struct DebugAndClone<T> {
    _: PhantomData<T>,
}
impl<T> DebugAndClone<T> {
    fn new() -> Self { Self { _: Default::default() } }

    fn debug(&self, value: &Box<dyn Any + Send + Sync>, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        fmt::Debug::fmt(value.downcast_ref::<T>().expect("type-checked"), f)
    }

    // ...
}
```
If it is, then this should compile down to zero overhead.

Reply:
I've considered ways to make it happen since I like the idea, but it seems that an obstacle here is a generic parameter `T`, e.g. if `DebugAndClone` is to be stored in `TypeErasedBox` the generic parameter is unfortunately propagated up to `TypeErasedBox`:
```rust
struct TypeErasedBox<T> {
    field: Box<dyn Any + Send + Sync>,
    debug_and_clone: DebugAndClone<T>,
}
```
This, in turn, makes `T` also appear in `Layer` because it contains `TypeErasedBox<T>`.

I assume we'd want to avoid all of this?

Reply:
https://github.com/awslabs/smithy-rs/issues/2794 created in case we come up with a different approach.